app/src/main/java/com/example/composetemplate/repositories/LoginRepository.kt

```diff
@@ -1,104 +1,188 @@
-                createOrUpdateUser(user, loginCallback)
-            else {
-                loginCallback(null, exception)
+    /**
+     * Registers a new user with the provided email and password.
+     *
+     * @param user The [User] object containing user details.
+     * @param password The password for the user.
+     * @return The result of the sign-up operation, which can either be success or failure.
+     */
+    suspend fun registerUser(user: User, password: String): SignUpResult {
+        return try {
+            when (val result = authActionable.createUserWithEmailAndPassword(user, password)) {
+                /* Handle cancellation or failure cases */
+                SignUpResult.Cancelled, is SignUpResult.Failure -> result
+                /* On successful creation, update the user in the local database */
+                is SignUpResult.Success -> {
+                    val success = createOrUpdateUser(user)
+                    if (success) result
+                    else SignUpResult.Cancelled
+                }
             }
+        } catch (e: Exception) {
+            /* Return failure result on exception */
+            SignUpResult.Failure(AuthError(e.errorType))
         }
     }
 
-    /** sign in user by email and password */
-    fun signInEmailPasswordUser(user: User, password: String, loginCallback: LoginCallback) {
-        if (user.email.isEmpty().not()) {
-            authActionable.signInWithEmailAndPassword(user, password) { createdUser, exception ->
-                if (createdUser != null && exception == null)
-                    loginCallback(user, null)
-                else {
-                    loginCallback(null, exception)
+    /**
+     * Signs in a user using the provided email and password.
+     *
+     * @param user The [User] object containing user details.
+     * @param password The password for the user.
+     * @return The result of the sign-in operation.
+     */
+    suspend fun signInEmailPasswordUser(user: User, password: String) = authActionable.signInWithEmailAndPassword(user, password)
+
+    /**
+     * Handles the Google Sign-In process and subsequent user management in the database.
+     *
+     * @param googleCredentialAuthParameter Parameters required for Google Sign-In.
+     * @return A [SignInResult] indicating the outcome of the sign-in and user management process.
+     */
+    suspend fun googleSignIn(googleCredentialAuthParameter: GoogleCredentialAuthParameter): SignInResult {
+        when (val signInResult = authActionable.googleSignIn(googleCredentialAuthParameter)) {
+            /* If the sign-in is cancelled, failed, or has no credentials, return that result */
+            SignInResult.Cancelled,
+            is SignInResult.Failure,
+            is SignInResult.NoCredentials -> return signInResult
+
+            /* If sign-in is successful, check if the user exists */
+            is SignInResult.Success -> {
```

Comment:
A. I would move the success logic to a separate function.
B. Is it possible to enter the Success case for googleSignIn but have getUser return Cancelled or Failure?
If the Google sign-in was successful, I would expect to already have a user.

Reply:
I will answer B. 
the getUser() call is to get the user from realtime database.  if it return nothing but success it mean there is no user in the database from some reason but it doesnt say that its not save in the authentication of firebase, so if authActionable.googleSignIn(googleCredentialAuthParameter) return success there is auth in firebase and we create it in the database in case it doesnt exist from some reason. 
read the comment I wrote.